dali/operators/math/expressions/broadcasting.cc

```diff
@@ -13,81 +13,117 @@
-  int ndim = shapes[0].sample_dim();
-  for (int i = 1; i < shapes.size(); i++) {
-    ndim = std::max(ndim, shapes[i].sample_dim());
+int BroadcastNdim(span<const TensorListShape<>*> shapes) {
+  return BroadcastNdimImpl(shapes);
+}
+
+/**
+ * @brief Implementation helper for `BroadcastShape` and `CanBroadcast`
+ *        The function updates val with the d-th dimension of `sh`, or returns false.
+ *        `val` can be updated with a new value if its current value is 1 or equal to the new one.
+ * @tparam ShapeLike e.g. span<const int64_t>, TensorShape<>
+ * @param val extent value to update (in an output shape)
+ * @param sh one of input shape
+ * @param d one of the dimension indices
+ * @return true if val could be updated (broadcastable)
+ * @return false if val could not be updated (non broadcastable)
+ */
+template <typename ShapeLike>
+bool BroadcastShapeImpl(int64_t& val, const ShapeLike& sh, int d) {
+  int ndim = sh.size();
+  auto extent = d < ndim ? sh[ndim - 1 - d] : 1;
+  if (extent > 1 && extent != val) {
+    if (val > 1)
+      return false;
+    val = extent;
   }
-  return ndim;
+  return true;
 }
 
-template <typename Shape>
-void BroadcastShapeImpl(Shape& result, span<const Shape> shapes) {
+std::string PrintShape(const TensorShape<>& shape, int sample_idx, int d) {
+  (void) sample_idx;
+  return make_string(to_string(shape), "(d=", shape.sample_dim() - d - 1, ")");
+}
+
+std::string PrintShape(const TensorListShape<>& shape, int sample_idx, int d) {
+  return make_string(to_string(shape[sample_idx]), " (d=", shape.sample_dim() - d - 1,
+                     ", belonging to sample_idx=", sample_idx, ")");
+}
+
+template <typename Shapes>
+std::string BroadcastErrorMessage(const Shapes& shapes, int64_t res,
+                                  int sample_idx, int d) {
+  std::stringstream ss;
+  ss << "Can't broadcast " << d << "-th dimension of shapes: ";
+  for (int j = 0; j < shapes.size(); j++) {
+    ss << "\n" << PrintShape(*shapes[j], sample_idx, d);
+    if (j == shapes.size() - 1)
+      ss << "\n";
+  }
+  return ss.str();
+}
+
+void BroadcastShape(TensorShape<>& result, span<const TensorShape<>*> shapes) {
   DALI_ENFORCE(shapes.size() >= 1);
   if (shapes.size() == 1) {
-    result = shapes[0];
+    result = *shapes[0];
     return;
   }
+
   int ndim = BroadcastNdim(shapes);
-  DALI_ENFORCE(result.size() == ndim);
+  result.resize(ndim);
 
   for (int d = 0; d < ndim; d++) {
-    auto &res = result[result.size() - 1 - d];
+    auto &res = result[ndim - 1 - d];
     res = 1;
     for (int i = 0; i < shapes.size(); i++) {
-      auto &sh = shapes[i];
-      auto extent = d < sh.size() ? sh[sh.size() - 1 - d] : 1;
-      if (extent > 1 && extent != res) {
-        if (res > 1)
-          DALI_FAIL(make_string("Can't broadcast extents ", extent, " and ", res));
-        res = extent;
+      if (!BroadcastShapeImpl(res, *shapes[i], d)) {
+        DALI_FAIL(BroadcastErrorMessage(shapes, res, i, d));
```

Comment:
We mix the sample_idx in BroadcastErrorMessage with the `i` that stands for the argument index basically.